tools/defaults.bzl

```diff
@@ -79,6 +95,73 @@ def pkg_npm(name, use_prodmode_output = False, **kwargs):
+    # prepare it for release.
+    jq(
+        name = "basic_substitutions",
+        srcs = ["//:package.json", pkg_json],
+        filter = """
+            # Root package.json
+            .[0] as $root
+            # Project package.json
+            | .[1] as $proj
+            # Get the fields from root package.json that should override the project
+            # package.json, i.e., every field except the following
+            | ($root
+                | del(.bin, .description, .dependencies, .name, .main, .peerDependencies, .optionalDependencies, .typings, .version, .private, .workspaces, .resolutions, .scripts, .[\"ng-update\"])
+            ) as $root_overrides
+            # Use the project package.json as a base and override other fields from root
+            | $proj + $root_overrides
+            # Combine keywords from both
+            | .keywords = ($root.keywords + $proj.keywords | unique)
+            # Remove devDependencies
+            | del(.devDependencies)
+            # Add engines
+            + {\"engines\": {\"node\": \"^12.20.0 || ^14.15.0 || >=16.10.0\", \"npm\": \"^6.11.0 || ^7.5.6 || >=8.0.0\", \"yarn\": \">= 1.13.0\"}}""",
+        args = ["--slurp"],
+        out = "substituted/package.json",
+    )
+
+    # Copy package.json files to bazel-out so we can use their bazel-out paths to determine
+    # the corresponding package npm package tar.gz path for substitutions.
+    copy_to_bin(
+        name = "package_json_copy",
+        srcs = [pkg_json],
+    )
+    pkg_deps_copies = []
+    for pkg_dep in pkg_deps:
+        pkg_label = to_label(pkg_dep)
+        if pkg_label.name != "package.json":
+            fail("ERROR: only package.json files allowed in pkg_deps of pkg_npm macro")
+        pkg_deps_copies.append("@%s//%s:package_json_copy" % (pkg_label.workspace_name, pkg_label.package))
+
+    # Substitute dependency versions for other packages in this repo with tar paths.
+    substitute_tar_deps(
+        name = "tar_substitutions",
+        src = "substituted/package.json",
+        pkg_deps = [":package_json_copy"] + pkg_deps_copies,
+        out = "substituted_with_tars/package.json",
+    )
+
+    # Move the generated package.json along with other deps into a directory for pkg_npm
+    # to package up because pkg_npm requires that all inputs be in the same directory.
+    copy_to_directory(
+        name = "package",
+        srcs = select({
+            # Do tar substitution if config_setting 'package_json_use_tar_deps' is true (local builds)
+            "//:package_json_use_tar_deps": [":%s_js_module_output" % name, "substituted_with_tars/package.json"],
+            "//conditions:default": [":%s_js_module_output" % name, "substituted/package.json"],
+        }),
+        replace_prefixes = {
+            "substituted_with_tars/": "",
+            "substituted/": "",
+        },
+        exclude_prefixes = [
+            "node_modules",
```

Comment:
Are these `exclude_prefixes` still needed or just left over from earlier?

Reply:
Still needed.

Reply:
Cool. Can you comment why? Not clear at first glance

Reply:
I was mistaken. `node_modules` doesn't need to be excluded, but `packages` does. Otherwise the packages contain compiled outputs of the other packages.

Reply:
As discussed in standup, the `extract_js_module_output` includes transitive dependencies on other packages, so excluding `packages` is required. Excluding `node_modules` is no longer required due to your change to `copy_to_directory`.